image/tarexport/save_oci.go

```diff
@@ -0,0 +1,404 @@
+package tarexport
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"runtime"
+
+	"github.com/docker/distribution/reference"
+	"github.com/docker/docker/image"
+	"github.com/docker/docker/layer"
+	"github.com/docker/docker/pkg/archive"
+	"github.com/docker/docker/pkg/system"
+	"github.com/opencontainers/go-digest"
+	imgspec "github.com/opencontainers/image-spec/specs-go"
+	ociv1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+type layerInfo struct {
+	digest digest.Digest
+	size   int64
+}
+
+type ociSaveSession struct {
+	*tarexporter
+	images       map[image.ID]*imageDescriptor
+	name         string
+	savedImages  map[image.ID]ociv1.Manifest // cache image.ID -> manifest bytes
+	diffIDsCache map[layer.DiffID]*layerInfo
+	outDir       string
+	index        ociv1.Index
+}
+
+func (l *tarexporter) getRefs() (map[string]string, map[string]reference.NamedTagged, error) {
+	refs := make(map[string]string)
+	reversed := make(map[string]reference.NamedTagged)
+	for image, ref := range l.refs {
+		r, err := reference.ParseNormalizedNamed(image)
+		if err != nil {
+			return nil, nil, err
+		}
+		if _, ok := r.(reference.Canonical); ok {
+			continue // a digest reference it's unique, no need for a --ref
+		}
+		tagged, ok := reference.TagNameOnly(r).(reference.NamedTagged)
+		if !ok {
+			return nil, nil, fmt.Errorf("error adding default tag to ref %q", ref)
+		}
+		if !ociRefRegexp.MatchString(ref) {
+			return nil, nil, fmt.Errorf(`invalid reference "%s=%s", reference must not include characters outside of the set of "A" to "Z", "a" to "z", "0" to "9", the hyphen "-", the dot ".", and the underscore "_"`, image, ref)
+		}
+		refs[reference.FamiliarString(tagged)] = ref
+		reversed[ref] = tagged
+	}
+	return refs, reversed, nil
+}
+
+var ociRefRegexp = regexp.MustCompile(`^([A-Za-z0-9._-]+)+$`)
+
+func (l *tarexporter) parseOCINames(names []string) (map[image.ID]*imageDescriptor, error) {
+	refs, _, err := l.getRefs()
+	if err != nil {
+		return nil, err
+	}
+	imgDescr := make(map[image.ID]*imageDescriptor)
+	tags := make(map[string]bool)
+
+	addAssoc := func(id image.ID, ref reference.Named) error {
+		if _, ok := imgDescr[id]; !ok {
+			imgDescr[id] = &imageDescriptor{}
+		}
+
+		if ref != nil {
+			if _, ok := ref.(reference.Canonical); ok {
+				return nil
+			}
+			tagged, ok := reference.TagNameOnly(ref).(reference.NamedTagged)
+			if !ok {
+				return nil
+			}
+
+			r, ok := refs[reference.FamiliarString(tagged)]
+			if ok {
+				var err error
+				if tagged, err = reference.WithTag(tagged, r); err != nil {
+					return err
+				}
+			}
+
+			for _, t := range imgDescr[id].refs {
+				if reference.FamiliarString(tagged) == reference.FamiliarString(t) {
+					return nil
+				}
+			}
+
+			if tags[tagged.Tag()] {
+				return fmt.Errorf("unable to include unique references %q in OCI image", tagged.Tag())
+			}
+
+			tags[tagged.Tag()] = true
```

Comment:
So you can't save two different images, like:
`docker image save ubuntu:latest busybox:latest -o image-latest-save.tar` 
(which works today)?